File: 0671-second-minimum-node-in-a-binary-tree/0671-second-minimum-node-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int findSecondMinimumValue(TreeNode root) {
        if(root==null) return -1;
        
        if(root.left == null && root.right == null) return -1;
        
        int leftMin = root.left.val;
        int rightMin = root.right.val;
        
        if(root.val == leftMin) leftMin = findSecondMinimumValue(root.left);
        if(root.val == rightMin) rightMin = findSecondMinimumValue(root.right);
        
        if(leftMin != -1 && rightMin != -1) {
            return Math.min(leftMin,rightMin);
        }
        else if(leftMin != -1){
            return leftMin;
        }
        else return rightMin;
    }
}

File: 0671-second-minimum-node-in-a-binary-tree/0671-second-minimum-node-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        //Base case for null root
        if(!root) return -1;
        //Base Case for leaf nodes
        if(!root->left && !root->right) return -1;
        
        int leftSMin = root->left->val;
        int rightSMin = root->right->val;
        
        //Conditional Recursive Call
        if(root->val == root->left->val) leftSMin = findSecondMinimumValue(root->left);
        
        if(root->val == root->right->val) rightSMin = findSecondMinimumValue(root->right);
        
        if(leftSMin != -1 && rightSMin != -1){
            return min(leftSMin,rightSMin);
        }
        else if(leftSMin != -1) {
            return leftSMin;
        }
        else return rightSMin;
    }
};